epub33/core/index.html

```diff
@@ -5301,11 +5332,22 @@ <h4>File and Directory Structure</h4>
 				</section>
 
 				<section id="sec-container-iri">
-					<h4>Relative URLs for Referencing Other Components</h4>
+					<h4>URLs in the OCF Abstract Container</h4>
 
-					<p>Files within the <a>OCF Abstract Container</a> MUST reference each other via <a
-							href="https://url.spec.whatwg.org/#relative-url-with-fragment-string"
-							>relative-URL-with-fragment strings</a> [[URL]].</p>
+					<p>The <a>container root URL</a> is the <a data-cite="url#concept-url">URL</a> [[URL]] of the
+						<a>Root Directory</a>. It is implementation specific, but MUST abide to the following requirements:</p>
+
+					<ul>
+						<li>The result of <a data-cite="url#concept-url-parser">parsing </a> "<code>/</code>" with the <a>container root URL</a> as <a data-cite="url#concept-base-url">base</a> MUST be the <a>container root URL</a>.</li>
+						<li>The result of <a data-cite="url#concept-url-parser">parsing </a> "<code>..</code>" with the <a>container root URL</a> as <a data-cite="url#concept-base-url">base</a> MUST be the <a>container root URL</a>.</li>	
```

Comment:
I think I see what you are trying to say here, but it also seems overly specific. For instance, given these, what is the required result of parsing `../..` with the root as base? Maybe more generally what we want is "The result of parsing any relative URL with the container root URL as base MUST begin with the Container Root URL".

Reply:
> it also seems overly specific

The requirements are very specific, but are sufficient to ensure that it covers all the cases (if I'm not mistaken). It's a minimal characterization.

> what is the required result of parsing ../.. with the root as base?

If `parse('..',root) == root`, then `parse('../..', root) == root`, necessarily

Reply:
I see no reason why that follows from what is defined in the spec. That is, nowhere do we say that `parse("x/y", root)` is the equivalent `parse("y", parse("x", root))`. If `parse('...', root)` is never a step in `parse('../..', root)` then I don't see how the restriction on the first applies to the second.

Reply:
> I see no reason why that follows from what is defined in the spec.

It's not directly defined in the spec, but it follows from the (normatively referenced) [URL parsing algorithm](https://url.spec.whatwg.org/#concept-url-parser).

Reply:
@bduga are you o.k. with this explanation (ie, can this be considered as resolved?)

Reply:
Sorry I meant to comment on it yesterday, but got distracted. This doesn't really address my point, and in fact I am not really sure what the intended behavior is. I don't understand from this statement where this alteration occurs during the parsing.I see two possibilities: 

1. When the _input_ to the parse algorithm is the string `..`
2. Whenever a _double-dot path segment_ is encountered while parsing, `shorten` should not happen to the path if the current path is the root

Option 2 seems implausible for us to really define, except to say broadly that any result of parsing that doesn't start with the root should be replaced with the root. Option 1 is easier, but less useful since the string `../..` is not identical to `..`. Option 1 also seems to me to be what the current text says. I guess my question is, what do we expect when parsing `../OPS` with a root of `MyRoot/`? Is it `MyRoot/` or `MyRoot/OPS`? Or even `OPS`, which is what I think the current spec text says.

Reply:
> in fact I am not really sure what the intended behavior is. I don't understand from this statement where this alteration occurs during the parsing

The intention is not to _alter_ the parsing algorithm, but to put a requirement on the Root URL such that the parsing becomes predictable.

Consider this statement:

> **if** parsing `".."` with _`root`_ is _`root`_, **then** parsing `"../.."` with _`root`_ is _`root`_.

It seems you do not agree that this statement is true. I'm saying it is, by virtue of following the URL parsing algorithm.

Why does it matter? To make the result of parsing predictable.

Let's say I have the following container structure:

```
/
├── mimetype
├── META-INF
│   └── container.xml
└── EPUB
    ├── package.opf
    ├── doc.xhtml
    └── …
```

With the Package Document containing the following item:

```
<item href="../EPUB/doc.xhtml"/>
```

The constraint is critical to say that this EPUB is conforming, as the result of parsing `"../EPUB/doc.xhtml"` is the same as the result of parsing "EPUB/doc.xhtml`, which is the URL of a file in the container.

  * If _Acme RS_'s implementation defines the root URL as being `"http://localhost:42424/"`, the result of parsing the item's URL is `"http://localhost:42424/EPUB/doc.xhtml"`.
  * If _Melville RS_'s implementation defines the root URL as being `"file:/MyRoot/"`, the result of parsing the item's URL is `"file:/EPUB/doc.xhtml"`.

_Melville RS_’s implementation is "leaky" (a parsed URL can represent something outside the root).

With the proposed requirement:
  * _Melville RS_ becomes non-conforming
  * we're sure that the result of parsing a URL string does represent something inside the container root (even if an actual resource doesn't exist there).
  * we're sure that a URL string that parses to the URL of an existing file in one implementation will always parse to the URL of the same existing file in other conforming implementations.

I hope this clarifies your concerns?

Reply:
@bduga, to answer your question more directly:
> I guess my question is, what do we expect when parsing `../OPS` with a root of `MyRoot/`? Is it `MyRoot/` or `MyRoot/OPS`? Or even `OPS`, which is what I think the current spec text says.

It's a bit difficult to say since `MyRoot/` is not a URL (it's a valid URL string, being a valid relative URL, but not a serialization of a full URL record).

But, assuming root URL is `http://localhost/MyRoot/`, the result of parsing `../OPS` is indeed `http://localhost/OPS`/ (so, as a root-relative path, that is `OPS`).

That's why `http://localhost/MyRoot/` is not conforming to my proposal.

Reply:
Ok, I think I finally see what you are saying. It is not that the RS must make sure to do these things, but rather that whatever implementation the RS chooses must have these properties. That is we are not requiring the RS to change parsing, we are saying the way the RS represents the container root URL cause the defined parsing algorithm to behave in this manner. Taking a quick stab at clarifying, maybe something like this:

>The container root URL is the URL [[URL]] of the Root Directory. It is implementation specific, but the implementation MUST have the following properties:
>
>	The result of parsing "/" with the container root URL as base is the container root URL.
>	The result of parsing ".." with the container root URL as base is the container root URL.

I still think this may be awkward, but at least it highlights the fact that these are properties of the implementation and not changes to parsing. I also remove the musts from the elements to make it clear that the RS is not expected to do anything itself in these cases, the must only applies to the implementation of the root. It's also strange to must musts.

Reply:
> It is not that the RS must make sure to do these things, but rather that whatever implementation the RS chooses must have these properties. That is we are not requiring the RS to change parsing, we are saying the way the RS represents the container root URL cause the defined parsing algorithm to behave in this manner.

Exactly!

Your suggested rewording also works for me. I'm all for not musting musts 😁.

Reply:
@bduga @rdeltour I have made the editorial change that you have converged to (will come with the next commit). Just to be on the safe side, I do not push the 'Resolve conversation' button, I leave that to you guys if you feel this is o.k.

Reply:
I'm OK to resolve this conversation. The change is editorial but a little clearer.
I'll let @bduga have the last word on it 😊